librad/tests/propagation_basic.rs

```diff
@@ -36,104 +37,124 @@ use fixtures::{Alice, Radicle};
+                    .unwrap();
+                // sanity check
+                git2.open_repo(radicle.urn()).unwrap();
+            }
         })
-        .await;
-
-        urn
-    };
-
-    let git1 = peers[1].peer.git().clone();
-    let _ = git1.open_repo(urn).unwrap();
+        .await
+        .unwrap();
+    })
+    .await;
 }
 
 #[tokio::test]
 async fn fetches_on_gossip_notify() {
     logging::init();
 
-    let peers = testnet::setup(2).unwrap();
-    let bound = testnet::bind(&peers).await.unwrap();
-
-    let urn = {
-        let peer1 = peers[0].peer.clone();
-        let peer2 = peers[1].peer.clone();
+    let peers = testnet::setup(2).await.unwrap();
+    testnet::run_on_testnet(peers, async move |mut apis| {
+        let peer1 = apis.pop().unwrap();
+        let peer2 = apis.pop().unwrap();
 
         let alice = Alice::new(peer1.public_key());
         let mut radicle = Radicle::new(&alice);
-        let urn = radicle.urn();
+        radicle
+            .sign(peer1.key(), &Signatory::User(alice.urn()), &alice)
+            .await
+            .unwrap();
 
-        run_on_testnet(bound, async move {
-            radicle
-                .sign(peer1.key(), &Signatory::User(alice.urn()), &alice)
-                .await
-                .unwrap();
+        let peer1_storage = peer1.storage();
+        let peer2_storage = peer2.storage();
 
+        // Create project on peer1, and clone from peer2
+        {
+            peer1_storage.create_repo(&radicle).unwrap();
+            let peer2_storage = peer2_storage.reopen().unwrap();
+            let url = radicle.urn().into_rad_url(peer1.peer_id());
             tokio::task::spawn_blocking(move || {
-                let git1 = peer1.git().clone();
-                let git2 = peer2.git().clone();
-                git1.create_repo(&radicle).unwrap();
-                git2.clone_repo::<ProjectInfo>(
-                    radicle.urn().into_rad_url(PeerId::from(peer1.public_key())),
-                )
-                .unwrap();
+                peer2_storage.clone_repo::<ProjectInfo>(url).unwrap();
             })
             .await
-            .unwrap()
-        })
-        .await;
-
-        urn
-    };
-
-    let git1 = peers[1].peer.git().clone();
-    let _ = git1.open_repo(urn).unwrap();
-}
-
-async fn run_on_testnet<F, A>(bound: Vec<BoundPeer<'_>>, f: F) -> A
-where
-    F: Future<Output = A>,
-{
-    let handles = bound.iter().map(|b| b.handle()).collect();
-    let connected = testnet::wait_connected(handles, bound.len());
-
-    let res = futures::future::select(
-        Box::pin(testnet::run(bound, Monitor::new())),
-        Box::pin(async {
-            connected.await;
-            f.await
-        }),
-    )
+            .unwrap();
+        }
+
+        // Add a commit on peer1
+        let commit_id = {
+            let repo = peer1_storage.open_repo(radicle.urn()).unwrap();
+            let empty_tree = {
+                let mut index = repo.index().unwrap();
```

Comment:
Note that this one is pretty unsound -- the index is not meant for concurrent access, let alone on a bare repo, afaik. In reality, these commits will be added in the working copy, though.